### PR TITLE
LibC: Flush all file streams on exit

### PR DESCRIPTION
**Pull request by BertalanD**
This is another step on my journey to getting `bash` to build inside Serenity.

---
The POSIX standard specifies the following:

> If the main() function returns to its original caller, or if the
> exit() function is called, all open files are closed (hence all output
> streams are flushed) before program termination

This means that flushing `stdin` and `stdout` only is not enough, as the
program might have pending writes in other file buffers too.

We now keep a linked list of all open file streams, and flush them on
program termination. This fixes one of bash's generated headers not
being written to disk.

### CODE REVIEW

File: Userland/Libraries/LibC/stdio.cpp [hunk 1]
```diff
@@ -591,8 +591,14 @@ int feof(FILE* stream)
 int fflush(FILE* stream)
 {
     if (!stream) {
-        dbgln("FIXME: fflush(nullptr) should flush all open streams");
-        return 0;
+        int rc = 0;
+        __pthread_mutex_lock(&s_open_streams_lock);
+        for (auto& file : *s_open_streams) {
+            ScopedFileLock lock(&file);
+            rc = file.flush() ? rc : EOF;
```
Comment (Quaker762) [inline Userland/Libraries/LibC/stdio.cpp:598]: Do we need to set the `EOF` indicator of the stream here as well, or is this done for free inside `::flush()`?

Comment (Quaker762) [reply Userland/Libraries/LibC/stdio.cpp:598]: This is already done here: https://github.com/SerenityOS/serenity/blob/6e37487477ae3059350a32baa337e376337ad752/Userland/Libraries/LibC/stdio.cpp#L103

Comment (BertalanD) [reply Userland/Libraries/LibC/stdio.cpp:598]: The error code is set inside `FILE::do_write()`.

File: Userland/Libraries/LibC/stdio.cpp [hunk 2]
```diff
@@ -1070,6 +1095,9 @@ int fclose(FILE* stream)
     }
     ScopedValueRollback errno_restorer(errno);
 
+    __pthread_mutex_lock(&s_open_streams_lock);
```
Comment (bgianfo) [inline Userland/Libraries/LibC/stdio.cpp:1098]: Can we add an RAII pthread locker type to avoid these manual locks unlocks?

**Comment by bgianfo** [comment]
Are there any meaningful tests we could write for this?

**Pull request merged**